packages/react-redux-subspace/src/hooks/useSubspace.js

```diff
@@ -6,31 +6,40 @@
  * LICENSE file in the root directory of this source tree.
  */
 
-import { useContext, useMemo } from "react"
+import { useContext, useMemo, useRef } from "react"
 import { ReactReduxContext } from "react-redux"
 import { subspace } from "redux-subspace"
-import isObject from '../utils/isObject'
+import isObject from "../utils/isObject"
 
-export const useSubspaceAdvanced = (
+export const useSubspace = (
   mapState,
   namespace,
-  subspaceDecorator,
-  { context = ReactReduxContext } = {}
+  options
 ) => {
-  const subspacer = useMemo(
-    () => subspaceDecorator || subspace(mapState, namespace),
-    [mapState, namespace]
-  )
-  const { store } = useContext(context)
-  return useMemo(() => subspacer(store), [subspacer, store])
-}
-
-const useSubspace = (mapState, namespace, options) => {
   if (options === undefined && isObject(namespace)) {
     return useSubspace(mapState, undefined, namespace)
   }
 
-  return useSubspaceAdvanced(mapState, namespace, undefined, options)
+  const { context = ReactReduxContext } = options || {}
+
+  const mapStateRef = useRef(mapState)
+  mapStateRef.current = mapState
+
+  const isMapStateFunction = typeof mapStateRef.current === "function"
+
+  const subspacer = useMemo(() => {
+    const mapStateHandler = isMapStateFunction
+      ? (state, rootState) => mapStateRef.current(state, rootState)
+      : mapStateRef.current
+
+    return subspace(mapStateHandler, namespace)
+  }, [
+    isMapStateFunction ? mapStateRef : mapStateRef.current,
+    namespace
+  ])
```

Comment:
Isn't that basically the same thing as the following?

```javascript
const isMapStateFunction = typeof mapStateRef.current === "function";
let subspacer;

if (isMapStateFunction) {
   subspacer = useMemo(() => subspace(mapState, namespace), [namespace])
} else {
  subspacer = useMemo(() => subspace(mapState, namespace), [
    mapState,
    namespace
  ]);
}
```

In case we're passing a function, compute the subspacer only when the namespace changes, otherwise compute it when the mapState object or namespace changes.

Reply:
Not quite.  This would fail to allow `mapState` function from being changed if the `namespace` did not.

I get what you're saying generally with using `isMapStateFunction` once and switching out which `useMemo` get's used.  The main reason I have a single `useMemo` was that I was trying hard not to violate the [one of the rules of hooks](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level).  I know that only a single `useMemo` would be called and one would be called every render, so I guess technically it wouldn't violate the rule, but I think the lint rule might flag it regardless, which will make others question it when they come across it.  

I'm also not sure how React deals with the dependency array changing length between renders.  It might be fine with it, but it also might do something weird.  I'd have to run a few experiments to check.

Assuming it all works,  I could see it becoming something like:

```js
const mapStateRef = useRef(mapState)
mapStateRef.current = mapState

const subspacer = typeof mapStateRef.current === "function"
  ? useMemo(() => subspace((...args) => mapStateRef.current(...args), namespace), [namespace])
  : useMemo(() => subspace(mapStateRef.current, namespace), [mapStateRef.current, namespace])
```

Reply:
Note, the changing dependencies length does produce a warning out of React.  Changing the array to include `mapStateRef` satisfies this requirement.

Reply:
huh, that's interesting;
my whole point was just mostly about - it's not immediately obvious what's happening there

Reply:
Do you find the new changes easier?

Reply:
Yeah.